packages/core/src/types.ts

```diff
@@ -0,0 +1,214 @@
+  // WalletConnect app deeplink formats, with {{uri}} replaced with the
+  // connection URI.
+  walletConnectDeeplinkFormat?: {
+    ios: string
+    android: string
+  }
+  // WalletConnect client signing methods.
+  walletConnectSigningMethods?: string[]
+  // A function that returns an instantiated wallet client, with `walletConnect`
+  // and `newWalletConnectSession` passed if `isWalletConnect === true`.
+  getClient: (
+    chainInfo: ChainInfo,
+    walletConnect?: WalletConnect,
+    newWalletConnectSession?: boolean
+  ) => Promise<Client | undefined>
+  // A function that returns the function to retrieve the `OfflineSigner` for
+  // this wallet.
+  getOfflineSignerFunction: (
+    client: Client
+  ) => (chainId: string) => OfflineSigner | Promise<OfflineSigner>
+  // A function that enables the client.
+  enableClient: (client: Client, chainInfo: ChainInfo) => Promise<void>
+  // A function that is called after a connection attempt completes. Will fail
+  // silently if an error is thrown.
+  cleanupClient?: (client: Client) => Promise<void>
+  // A function that returns the wallet name and address from the client.
+  getNameAddress: (
+    client: Client,
+    chainInfo: ChainInfo
+  ) => Promise<{ name: string; address: string }>
+  // A function that determines if this wallet should automatically be connected
+  // on initialization.
+  shouldAutoconnect?: () => boolean | Promise<boolean>
+  // A function that will execute the passed listener when the wallet connection
+  // data needs to be refreshed. This will likely be used when the user switches
+  // accounts in the wallet, and the name and address need to be updated. Called
+  // on successful wallet connection.
+  addRefreshListener?: (listener: () => void) => void
+  // A function that will remove the refresh listener added previously. Called
+  // on wallet disconnect.
+  removeRefreshListener?: (listener: () => void) => void
+}
+
+export interface ConnectedWallet<Client = unknown> {
+  // Wallet.
+  wallet: Wallet<Client>
+  // Wallet client.
+  walletClient: Client
+  // Chain info the clients are connected to.
+  chainInfo: ChainInfo
+  // Offline signer for the wallet client.
+  offlineSigner: OfflineSigner
+  // User's name for their wallet.
+  name: string
+  // Wallet address.
+  address: string
+  // Signing client for interacting with CosmWasm chain APIs.
+  signingCosmWasmClient: SigningCosmWasmClient
+  // Signing client for interacting with Stargate chain APIs.
+  signingStargateClient: SigningStargateClient
+}
+
+export type SigningClientGetter<T> = (
+  chainInfo: ChainInfo
+) => T | Promise<T | undefined> | undefined
+
+export type ChainInfoOverrides =
+  | ChainInfo[]
+  | (() => undefined | ChainInfo[] | Promise<undefined | ChainInfo[]>)
+
+export enum CosmosWalletStatus {
+  Uninitialized,
+  // Don't call connect until this state is reached.
+  Disconnected,
+  Connecting,
+  Connected,
+  Errored,
+}
+
+export enum ChainInfoID {
```

Comment:
re: chain-registry — don't feel the need to use this yet, I'll add this stuff to the chain-registry issue.

Reply:
Yeah, we should definitely leverage the registry somehow, maybe soon after merge. I really dislike allowing arbitrary strings as parameters when there is only a finite set of options, which is why I created the `ChainInfoID` enum. This lets them neatly/readably set `defaultChainId` and override chain infos using the enum (in the `react` package), with confidence that it will be valid, preventing any typos and other careless mistakes.

Reply:
oh true, ok!